File: ARR/backend/land/services/pnu_resolver.py
"""
PNU Resolver - validates/parses 19-digit PNU codes and resolves addresses via Vworld API.

PNU structure (19 digits):
  [시도2][시군구3][읍면동3][리2][토지구분1][본번4][부번4] 토지구분: 1=대, 2=임야

Vworld Geocoding:
  1) 주소 → 좌표 (api.vworld.kr/req/address)
  2) 좌표 → 지번주소 (api.vworld.kr/req/address reverse)
  Requires VWORLD_API_KEY env var (https://www.vworld.kr 에서 발급)
"""

import logging
import os
import re

import httpx

logger = logging.getLogger(__name__)

_PNU_PATTERN = re.compile(r"^\d{19}$")
_TIMEOUT = httpx.Timeout(10.0, connect=5.0)

VWORLD_API_KEY = os.getenv("VWORLD_API_KEY", "")
VWORLD_GEOCODE_URL = "https://api.vworld.kr/req/address"


def validate_pnu(pnu: str) -> bool:
    """Check if a string is a valid 19-digit PNU code."""
    return bool(_PNU_PATTERN.match(pnu))


def parse_pnu(pnu: str) -> dict | None:
    """
    Parse a 19-digit PNU into components.

    Returns None if invalid, otherwise:
    {
        "pnu": str,
        "sido": str (2),
        "sigungu": str (3),
        "eupmyeondong": str (3),
        "ri": str (2),
        "land_type": str (1) - "1"=대, "2"=임야,
        "main_number": str (4),
        "sub_number": str (4)
    }
    """
    if not validate_pnu(pnu):
        return None

    return {
        "pnu": pnu,
        "sido": pnu[0:2],
        "sigungu": pnu[2:5],
        "eupmyeondong": pnu[5:8],
        "ri": pnu[8:10],
        "land_type": pnu[10:11],
        "main_number": pnu[11:15],
        "sub_number": pnu[15:19],
        "land_type_name": "대" if pnu[10] == "1" else "임야" if pnu[10] == "2" else "기타",
    }


def resolve_address(address: str) -> dict:
    """
    Resolve a Korean address to PNU via Vworld Geocoding API.

    Flow: 주소 → Vworld geocode → 좌표 + 법정동코드 + 지번

    Requires VWORLD_API_KEY env var.

    Returns:
        {
            "success": bool,
            "address": str,
            "pnu": str | None,
            "coordinates": {"x": float, "y": float} | None,
            "error": str (on failure)
        }
    """
    if not VWORLD_API_KEY:
        return {
            "success": False,
            "error": "VWORLD_API_KEY not configured. "
                     "Get a key at https://www.vworld.kr and set VWORLD_API_KEY env var.",
            "address": address,
            "pnu": None,
        }

    # Step 1: Geocode address (try PARCEL type for 지번, fallback to ROAD)
    for addr_type in ("PARCEL", "ROAD"):
        result = _vworld_geocode(address, addr_type)
        if result and result.get("response", {}).get("status") == "OK":
            return _parse_geocode_result(result, address)

    return {
        "success": False,
        "error": f"Vworld geocoding failed: no results for '{address}'",
        "address": address,
        "pnu": None,
    }


def _vworld_geocode(address: str, addr_type: str) -> dict | None:
    """Call Vworld geocoding API."""
    params = {
        "service": "address",
        "request": "getCoord",
        "key": VWORLD_API_KEY,
        "address": address,
        "type": addr_type,
        "format": "json",
        "crs": "EPSG:4326",
        "refine": "true",
    }

    try:
        with httpx.Client(timeout=_TIMEOUT) as client:
            resp = client.get(VWORLD_GEOCODE_URL, params=params)
            resp.raise_for_status()
            return resp.json()
    except httpx.ConnectError:
        logger.error("Vworld API unreachable")
        return None
    except Exception as e:
        logger.error(f"Vworld geocode failed: {e}")
        return None


def _parse_geocode_result(data: dict, original_address: str) -> dict:
    """Parse Vworld geocode response into our format."""
    try:
        response = data["response"]
        if response.get("status") != "OK":
            return {
                "success": False,
                "error": f"Vworld status: {response.get('status')}",
                "address": original_address,
                "pnu": None,
            }

        result = response["result"]
        point = result["point"]
        x = float(point["x"])
        y = float(point["y"])

        # Extract PNU from refined.structure.level4LC (19-digit code)
        pnu = None
        refined = response.get("refined", {})
        structure = refined.get("structure", {})
        level4lc = structure.get("level4LC", "")
        if validate_pnu(level4lc):
            pnu = level4lc

        geocoded_text = refined.get("text", "")

        return {
            "success": True,
            "address": original_address,
            "pnu": pnu,
            "coordinates": {"x": x, "y": y},
            "geocoded_address": geocoded_text,
        }

    except (KeyError, TypeError, ValueError) as e:
        return {
            "success": False,
            "error": f"Failed to parse Vworld response: {e}",
            "address": original_address,
            "pnu": None,
        }

File: AG-light/worker/src/land/pnu-resolver.ts
/**
 * PNU Resolver — validates/parses 19-digit PNU codes, resolves addresses via Vworld API.
 * Port of: ARR/backend/land/services/pnu_resolver.py
 *
 * PNU structure (19 digits):
 *   [시도2][시군구3][읍면동3][리2][토지구분1][본번4][부번4]
 *   토지구분: 1=대, 2=임야
 */

const PNU_RE = /^\d{19}$/;

const VWORLD_GEOCODE_URL = "https://api.vworld.kr/req/address";

export function validatePnu(pnu: string): boolean {
  return PNU_RE.test(pnu);
}

export function parsePnu(pnu: string) {
  if (!validatePnu(pnu)) return null;

  const landTypeCode = pnu[10];
  return {
    pnu,
    sido: pnu.slice(0, 2),
    sigungu: pnu.slice(2, 5),
    eupmyeondong: pnu.slice(5, 8),
    ri: pnu.slice(8, 10),
    land_type: pnu.slice(10, 11),
    main_number: pnu.slice(11, 15),
    sub_number: pnu.slice(15, 19),
    land_type_name:
      landTypeCode === "1" ? "대" : landTypeCode === "2" ? "임야" : "기타",
  };
}

export type ResolveResult = {
  success: boolean;
  address: string;
  pnu: string | null;
  coordinates: { x: number; y: number } | null;
  geocoded_address?: string;
  error?: string;
};

export async function resolveAddress(
  address: string,
  vworldApiKey: string,
): Promise<ResolveResult> {
  if (!vworldApiKey) {
    return {
      success: false,
      error: "VWORLD_API_KEY not configured",
      address,
      pnu: null,
      coordinates: null,
    };
  }

  // Try PARCEL first, then ROAD
  for (const addrType of ["PARCEL", "ROAD"] as const) {
    const result = await vworldGeocode(address, addrType, vworldApiKey);
    if (result?.response?.status === "OK") {
      return parseGeocodeResult(result, address);
    }
  }

  return {
    success: false,
    error: `Vworld geocoding failed: no results for '${address}'`,
    address,
    pnu: null,
    coordinates: null,
  };
}

async function vworldGeocode(
  address: string,
  addrType: string,
  apiKey: string,
): Promise<VworldResponse | null> {
  const params = new URLSearchParams({
    service: "address",
    request: "getCoord",
    key: apiKey,
    address,
    type: addrType,
    format: "json",
    crs: "EPSG:4326",
    refine: "true",
  });

  try {
    const resp = await fetch(`${VWORLD_GEOCODE_URL}?${params}`, {
      headers: { "Referer": "http://localhost", "Origin": "http://localhost" },
      signal: AbortSignal.timeout(10_000),
    });
    if (!resp.ok) return null;
    return (await resp.json()) as VworldResponse;
  } catch {
    return null;
  }
}

interface VworldResponse {
  response: {
    status: string;
    result?: { point: { x: string; y: string } };
    refined?: {
      text?: string;
      structure?: { level4LC?: string };
    };
  };
}

function parseGeocodeResult(
  data: VworldResponse,
  originalAddress: string,
): ResolveResult {
  try {
    const { response } = data;
    if (response.status !== "OK") {
      return {
        success: false,
        error: `Vworld status: ${response.status}`,
        address: originalAddress,
        pnu: null,
        coordinates: null,
      };
    }

    const point = response.result!.point;
    const x = parseFloat(point.x);
    const y = parseFloat(point.y);

    const level4lc = response.refined?.structure?.level4LC ?? "";
    const pnu = validatePnu(level4lc) ? level4lc : null;
    const geocodedText = response.refined?.text ?? "";

    return {
      success: true,
      address: originalAddress,
      pnu,
      coordinates: { x, y },
      geocoded_address: geocodedText,
    };
  } catch {
    return {
      success: false,
      error: "Failed to parse Vworld response",
      address: originalAddress,
      pnu: null,
      coordinates: null,
    };
  }
}
